common/src/main/java/noobanidus/mods/lootr/common/block/entity/BlockEntityTicker.java

```diff
@@ -13,187 +12,187 @@
-      // This check consults the actual chunk source.
-      if (!level.getChunkSource().hasChunk(entry.getPosition().getX() >> 4, entry.getPosition().getZ() >> 4)) {
+      ServerLevel level = server.getLevel(entry.dimension());
+      if (level == null) {
+        iterator.remove();
         continue;
       }
 
-      // And for all chunks in a 5x5 area. Modifying a block entity or block
-      // will cause it to propagate ticks, etc, which can cause chunk loading.
-      boolean skip = false;
-      for (ChunkPos chunkPos : entry.getChunkPositions()) {
-        if (!level.getChunkSource().hasChunk(chunkPos.x, chunkPos.z)) {
-          skip = true;
-          break;
+      switch (entry.getChunkLoadStatus(level)) {
+        case UNLOADED -> {
+          // the chunk has unloaded. this entry is no longer valid, and it will be added again if the chunk loads again.
+          iterator.remove();
+        }
+        case SURROUNDING_CHUNKS_NOT_LOADED -> {
+          // keep waiting for the surrounding chunks to load
+        }
+        case FULLY_LOADED -> {
+          replaceEntitiesInChunk(level, entry);
+          iterator.remove();
         }
       }
-      if (skip) {
-        continue;
-      }
+    }
 
-      // We also check against `HandleChunk`'s internal list of loaded chunks.
-      // Hence the duplication.
-      if (LootrAPI.anyUnloadedChunks(entry.getDimension(), entry.getChunkPositions())) {
-        continue;
-      }
+    synchronized (pendingEntries) {
+      blockEntityEntries.addAll(pendingEntries);
+      pendingEntries.clear();
+    }
+  }
 
-      if (level.getServer().getWorldData().worldGenOptions().generateStructures()) {
-        Registry<Structure> registry = level.registryAccess().registryOrThrow(Registries.STRUCTURE);
-        ChunkPos thisPos = new ChunkPos(entry.getPosition());
-        if (registry.getTag(LootrTags.Structure.STRUCTURE_BLACKLIST).filter(tag -> tag.size() != 0).isPresent()) {
-          if (LootrAPI.isTaggedStructurePresent(level, thisPos, LootrTags.Structure.STRUCTURE_BLACKLIST, entry.getPosition())) {
-            iterator.remove();
-            continue;
-          }
-        } else if (registry.getTag(LootrTags.Structure.STRUCTURE_WHITELIST).filter(tag -> tag.size() != 0)
-            .isPresent()) {
-          if (!LootrAPI.isTaggedStructurePresent(level, thisPos, LootrTags.Structure.STRUCTURE_WHITELIST, entry.getPosition())) {
-            iterator.remove();
-            continue;
-          }
-        }
-      }
+  private static boolean checkStructureValidity(ServerLevel level, ChunkPos chunkPos, BlockPos position) {
+    if (!level.getServer().getWorldData().worldGenOptions().generateStructures()) {
+      return true;
+    }
+    Registry<Structure> registry = level.registryAccess().registryOrThrow(Registries.STRUCTURE);
+    if (registry.getTag(LootrTags.Structure.STRUCTURE_BLACKLIST).filter(tag -> tag.size() != 0).isPresent()) {
```

Comment:
I'm also curious if caching these values (and clearing on tag update) would be helpful.